### DISCUSSION TITLE
Health watcher primary unreachable

### DISCUSSION THREAD

**Discussion by lynshi** [microsoft/CCF#3818] [General] [answered]
The [`healthy_watcher.py` script](https://github.com/microsoft/CCF/blob/a16d81be9680e6dd950c37381a4e264bff35d857/tests/infra/health_watcher.py#L79) currently returns `HealthState.unavailable` if the current primary is unreachable. Why is this determined to be an unrecoverable state instead of returning `HealthState.election` - wouldn't one of the nodes eventually start an election if it never hears from the primary whether or not the primary exists?

**Comment by jumaffre** [comment, marked as answer]
@lynshi: `HealthState.unavailable` does not indicate that the service is unrecoverable; it simply states that the current primary is not available and that an observer cannot tell for sure whether an election is in progress at this point in time. 

In fact, the `get_network_health()` functions doesn't decide if the service is unrecoverable but simply returns the current health state of the network. It is up to the [`run()` loop](https://github.com/microsoft/CCF/blob/a16d81be9680e6dd950c37381a4e264bff35d857/tests/infra/health_watcher.py#L146-L167) to decide whether the service is unrecoverable, which is declared as such if the service hasn't be `HealthState.stable` for more than `unstable_threshold_s` (which is `N` times the election timeout, with `N` to be decided by the operator based on how sensitive the health check should be).
